File: Day03_2015.py
x = 0
y = 0
sx = 0
sy = 0
rx = 0
ry = 0

visitDict = {}
visitDict[0, 0] = 1
uniqueVisits1 = 1
visitDict2 = {}
visitDict2[0, 0] = 1
uniqueVisits2 = 1
isSanta = True

with open("day03_2015.txt") as file:
    for line in file:
        for char in line:
            if char == '<':
                x -= 1
                if isSanta:
                    sx -= 1
                else:
                    rx -= 1
            elif char == '>':
                x += 1
                if isSanta:
                    sx += 1
                else:
                    rx += 1
            elif char == '^':
                y += 1
                if isSanta:
                    sy += 1
                else:
                    ry += 1
            elif char == 'v':
                y -= 1
                if isSanta:
                    sy -= 1
                else:
                    ry -= 1

            if (x, y) not in visitDict:
                visitDict[x, y] = 1
                uniqueVisits1 += 1

            if isSanta:
                if (sx, sy) not in visitDict2:
                    visitDict2[sx, sy] = 1
                    uniqueVisits2 += 1
            else:
                if (rx, ry) not in visitDict2:
                    visitDict2[rx, ry] = 1
                    uniqueVisits2 += 1
                    
            isSanta = not isSanta

file.close()

print("Number of unique houses visited by Santa alone: " + str(uniqueVisits1))
print("Number of unique houses visited by Santa and Robo-Santa: " + str(uniqueVisits2))

File: Day03_2015.cpp
#include <iostream>
#include <fstream>
#include <set>
#include <utility>

int main()
{
    int x, y, sx, sy, rx, ry;
    x = y = sx = sy = rx = ry = 0;

    std::set <std::pair <int, int>> visitMap1;
    std::set <std::pair <int, int>> visitMap2;
    visitMap1.emplace(0, 0);
    visitMap2.emplace(0, 0);

    int uniqueVisits1 = 1, uniqueVisits2 = 1;
    bool isSanta = true;

    std::ifstream file("Day03_2015.txt");
    char c;
    
    while (file.get(c))
    {
        if (c == '<')
        {
            x -= 1;
            if (isSanta)
                sx -= 1;
            else
                rx -= 1;
        }
        else if (c == '>')
        {
            x += 1;
            if (isSanta)
                sx += 1;
            else
                rx += 1;
        }
        else if (c == '^')
        {
            y += 1;
            if (isSanta)
                sy += 1;
            else
                ry += 1;
        }
        else if (c == 'v')
        {
            y -= 1;
            if (isSanta)
                sy -= 1;
            else
                ry -= 1;
        }

        if (visitMap1.count(std::make_pair(x, y)) == 0)
        {
            visitMap1.emplace(x, y);
            uniqueVisits1++;
        }
                
        if (isSanta)
        {
            if (visitMap2.count(std::make_pair(sx, sy)) == 0)
            {
                visitMap2.emplace(sx, sy);
                uniqueVisits2++;
            }
        }
        else
        {
            if (visitMap2.count(std::make_pair(rx, ry)) == 0)
            {
                visitMap2.emplace(rx, ry);
                uniqueVisits2++;
            }
        }

        isSanta = !isSanta;
    }

    file.close();

    std::cout << "Number of unique houses visited by Santa alone: " << uniqueVisits1 << "\n";
    std::cout << "Number of unique houses visited by Santa and Robo-Santa: " << uniqueVisits2;
}